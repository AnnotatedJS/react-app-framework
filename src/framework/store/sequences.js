import uniqueId from 'lodash/uniqueId';
import forOwn from 'lodash/forOwn';
import get from 'lodash/get';
import merge from 'lodash/merge';
import isArray from 'lodash/isArray';
import isObject from 'lodash/isObject';
import isEmpty from 'lodash/isEmpty';
import unionWith from 'lodash/unionWith';
import * as constants from './constants';

let userFunctions = {};

function getTargetPropertiesFromEvents(events, targetProperties) {
  if (events && events.length > 0) {
    events.forEach(event => {
      const { targets } = event;
      if (targets && targets.length > 0) {
        let key;
        let propertiesObject;
        targets.forEach(target => {
          const { type, props, events } = target;
          if (type === constants.COMPONENT_TYPE && props) {
            const { componentName, componentInstance, propertyName, forwardPath } = props;
            if (propertyName) {
              key = `${componentName}_${componentInstance}`;
              propertiesObject = targetProperties[key] || {};
              // tell the that this property should be bind to the http request query
              propertiesObject[propertyName] = merge({}, propertiesObject[propertyName], {
                forwardPath
              });
              targetProperties[key] = propertiesObject;
            }
          }
          getTargetPropertiesFromEvents(events, targetProperties);
        });
      }
    });
  }
}

function deriveTargetProperties(actionSequences, targetProperties = {}) {
  forOwn(actionSequences, (value, prop) => {
    getTargetPropertiesFromEvents(value.events, targetProperties);
  });
  return targetProperties;
}

function getEventSequence(event) {
  const eventSequence = {};
  const { name, targets } = event;
  if (targets && targets.length > 0) {
    eventSequence.name = name;
    eventSequence.targets = [];
    targets.forEach(target => {
      const { type, props, events } = target;
      if (props && !isEmpty(props)) {
        const newTarget = {
          type,
        };
        if (type === constants.USER_FUNCTION_TYPE) {
          newTarget.props = {
            ...props,
            functionKey: uniqueId('function')
          };
        } else if (type === constants.COMPONENT_TYPE) {
          newTarget.props = {
            ...props,
            componentKey: uniqueId('component')
          };
        }
        if (events && events.length > 0 && type === constants.USER_FUNCTION_TYPE) {
          const newTargetEvents = [];
          events.forEach(targetEvent => {
            newTargetEvents.push(getEventSequence(targetEvent));
          });
          newTarget.events = newTargetEvents;
        }
        eventSequence.targets.push(newTarget);
      }
    });
  }
  return eventSequence;
}

function eventTargetComparator(destTarget, sourceTarget) {
  const { type: sourceType, props: sourceProps } = sourceTarget;
  const { type: destType, props: destProps } = destTarget;
  if (sourceType === destType) {
    if (sourceProps && destProps) {
      if (sourceProps.functionName && destProps.functionName) {
        return sourceProps.functionName === destProps.functionName;
      } else if (sourceProps.componentName && destProps.componentName) {
        return sourceProps.componentName === destProps.componentName
          && sourceProps.componentInstance === destProps.componentInstance
          && sourceProps.propertyName === destProps.propertyName
          && sourceProps.forwardPath === destProps.forwardPath;
      }
    }
  }
  return false;
}

function targetEventComparator(destEvent, sourceEvent) {
  return destEvent === sourceEvent || destEvent.name === sourceEvent.name;
}

function mergeEventTargets(destTargets, sourceTargets) {
  if (destTargets !== sourceTargets) {
    let resultTargets = unionWith(destTargets, sourceTargets, eventTargetComparator);
    resultTargets.forEach(resultTarget => {
      const sameSourceTarget = sourceTargets.find(sourceTarget => eventTargetComparator(resultTarget, sourceTarget));
      if (sameSourceTarget) {
        const resultTargetEvents = unionWith(resultTarget.events, sameSourceTarget.events, targetEventComparator);
        if (resultTargetEvents && resultTargetEvents.length > 0) {
          resultTargetEvents.forEach(resultTargetEvent => {
            let sameSourceTargetEvent;
            if (sameSourceTarget.events && sameSourceTarget.events.length > 0) {
              sameSourceTargetEvent =
                sameSourceTarget.events.find(sourceTargetEvent => targetEventComparator(resultTargetEvent, sourceTargetEvent));
            } else if (resultTarget.events && resultTarget.events.length > 0) {
              sameSourceTargetEvent =
                resultTarget.events.find(sourceTargetEvent => targetEventComparator(resultTargetEvent, sourceTargetEvent));
            }
            if (sameSourceTargetEvent) {
              resultTargetEvent.targets = mergeEventTargets(resultTargetEvent.targets, sameSourceTargetEvent.targets);
            }
          });
        }
        resultTarget.events = resultTargetEvents;
      }
    });
    return resultTargets;
  }
  return destTargets;
}

function getActionSequences(handlers, actionSequences = {}) {
  if (handlers && handlers.length > 0) {
    handlers.forEach(handler => {
      const { type, props, events } = handler;
      if (events && events.length > 0) {
        events.forEach(event => {
          let key;
          let handlerObject;
          if (event && event.name && event.targets && event.targets.length > 0) {
            if (type === constants.COMPONENT_TYPE) {
              key = `${props.componentName}_${props.componentInstance}`;
              handlerObject = actionSequences[key]
                || { ...props, componentKey: uniqueId('component'), events: [] };
              const eventSequence = getEventSequence(event);
              // find the same event handler name for the container
              const existingHandlerEventIndex =
                handlerObject.events.findIndex(evn => evn.name === eventSequence.name);
              if (existingHandlerEventIndex >= 0) {
                // here we should merge targets of the same container events handler
                const existingHandlerEvent = handlerObject.events[existingHandlerEventIndex];
                if (existingHandlerEvent) {
                  handlerObject.events[existingHandlerEventIndex].targets =
                    mergeEventTargets(existingHandlerEvent.targets, eventSequence.targets);
                }
              } else {
                handlerObject.events.push(eventSequence);
              }
              actionSequences[key] = handlerObject;
            }
            getActionSequences(event.targets, actionSequences);
          }
        });
      }
    });
  }
}

function createActionSequencesRecursively (handlers, actionSequences = {}) {
  forOwn(handlers, value => {
    if (isArray(value)) {
      // only arrays should be exported as flow sequences otherwise the object is assumed as a nested sequences
      getActionSequences(value, actionSequences);
    } else if (isObject(value)) {
      // if the handlers is object - it means we have a nested handlers description
      createActionSequencesRecursively(value, actionSequences);
    }
  });
  return actionSequences;
}

export function createActionSequences (handlers, functions) {
  userFunctions = functions;
  const actionSequences = createActionSequencesRecursively(handlers);
  const targetProperties = deriveTargetProperties(actionSequences);
  return { actionSequences, targetProperties };
}

export function getUserFunctionByName (functionName) {
  return get(userFunctions, functionName);
}
