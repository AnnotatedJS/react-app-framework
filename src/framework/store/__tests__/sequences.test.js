import schema_empty from '../../../test/empty/app/schema';
import userFunctions_empty from '../../../test/empty/app/indices/userFunctions';
import schema_sequences from '../../../test/sequences/app/schema';
import userFunctions_sequences from '../../../test/sequences/app/indices/userFunctions';
import schema_merge from '../../../test/merge/app/schema';
import userFunctions_merge from '../../../test/merge/app/indices/userFunctions';

import { createActionSequences } from '../sequences';

describe('Sequences', () => {

  it('create action sequences with empty app files', () => {
    const { actionSequences, targetProperties } =
      createActionSequences(schema_empty.flows, userFunctions_empty);
    expect(actionSequences).toEqual({});
    expect(targetProperties).toEqual({});
  });

  it('create action sequences', () => {
    const { actionSequences, targetProperties } =
      createActionSequences(schema_sequences.flows, userFunctions_sequences);
    expect(actionSequences).toMatchSnapshot();
    expect(targetProperties).toMatchSnapshot();
  });

  it('merge sequences', () => {
      const { actionSequences, targetProperties } =
        createActionSequences(schema_merge.flows, userFunctions_merge);
    expect(actionSequences).toMatchSnapshot();
    expect(targetProperties).toMatchSnapshot();
  });

});