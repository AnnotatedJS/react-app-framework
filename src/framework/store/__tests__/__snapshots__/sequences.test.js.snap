// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequences create action sequences 1`] = `
Object {
  "usr.Form.Form_form": Object {
    "componentInstance": "form",
    "componentName": "usr.Form.Form",
    "events": Array [
      Object {
        "name": "onSubmit",
        "targets": Array [
          Object {
            "events": Array [],
            "props": Object {
              "componentInstance": "headerTest1",
              "componentName": "usr.Header.Header",
              "propertyName": "title",
            },
            "type": "component",
          },
          Object {
            "events": Array [
              Object {
                "name": "errors",
                "targets": Array [
                  Object {
                    "props": Object {
                      "componentInstance": "form",
                      "componentName": "usr.SubmitForm.SubmitForm",
                      "propertyName": "errors",
                    },
                    "type": "component",
                  },
                ],
              },
              Object {
                "name": "success",
                "targets": Array [
                  Object {
                    "events": Array [
                      Object {
                        "name": "loading",
                        "targets": Array [
                          Object {
                            "props": Object {
                              "componentInstance": "globalSpinner1",
                              "componentName": "usr.GlobalSpinner.GlobalSpinner",
                            },
                            "type": "component",
                          },
                        ],
                      },
                      Object {
                        "name": "loading",
                        "targets": Array [
                          Object {
                            "props": Object {
                              "componentInstance": "localSpinner1",
                              "componentName": "usr.LocalSpinner.LocalSpinner",
                            },
                            "type": "component",
                          },
                        ],
                      },
                      Object {
                        "name": "resultData",
                        "targets": Array [
                          Object {
                            "props": Object {
                              "componentInstance": "paneInfo",
                              "componentName": "usr.PaneInfo.PaneInfo",
                              "propertyName": "data",
                            },
                            "type": "component",
                          },
                        ],
                      },
                    ],
                    "props": Object {
                      "functionName": "usr.api.myFunctions.getDataByForm",
                    },
                    "type": "userFunction",
                  },
                ],
              },
            ],
            "props": Object {
              "functionName": "usr.api.myFunctions.checkFormValidity",
            },
            "type": "userFunction",
          },
        ],
      },
      Object {
        "name": "onCancel",
        "targets": Array [
          Object {
            "props": Object {
              "componentInstance": "headerTest1",
              "componentName": "usr.Header.Header",
              "propertyName": "title",
            },
            "type": "component",
          },
        ],
      },
      Object {
        "name": "onGoBack",
        "targets": Array [
          Object {
            "props": Object {
              "componentInstance": "landingPane",
              "componentName": "usr.Landing.Landing",
              "forwardPath": "landingPage",
              "propertyName": "formHistory",
            },
            "type": "component",
          },
          Object {
            "events": Array [],
            "props": Object {
              "componentInstance": "landingPane2",
              "componentName": "usr.Landing.Landing",
              "propertyName": "formHistory",
            },
            "type": "component",
          },
        ],
      },
    ],
  },
  "usr.SubmitForm.SubmitForm_form": Object {
    "componentInstance": "form",
    "componentName": "usr.SubmitForm.SubmitForm",
    "events": Array [
      Object {
        "name": "onCancel",
        "targets": Array [
          Object {
            "props": Object {
              "componentInstance": "headerTest",
              "componentName": "usr.Header.Header",
              "propertyName": "errorText",
            },
            "type": "component",
          },
        ],
      },
    ],
    "propertyName": "errors",
  },
}
`;

exports[`Sequences create action sequences 2`] = `
Object {
  "usr.Header.Header_headerTest": Object {
    "errorText": Object {
      "forwardPath": undefined,
    },
  },
  "usr.Header.Header_headerTest1": Object {
    "title": Object {
      "forwardPath": undefined,
    },
  },
  "usr.Landing.Landing_landingPane": Object {
    "formHistory": Object {
      "forwardPath": "landingPage",
    },
  },
  "usr.Landing.Landing_landingPane2": Object {
    "formHistory": Object {
      "forwardPath": undefined,
    },
  },
  "usr.PaneInfo.PaneInfo_paneInfo": Object {
    "data": Object {
      "forwardPath": undefined,
    },
  },
  "usr.SubmitForm.SubmitForm_form": Object {
    "errors": Object {
      "forwardPath": undefined,
    },
  },
}
`;

exports[`Sequences merge sequences 1`] = `
Object {
  "usr.Form.Form_form": Object {
    "componentInstance": "form",
    "componentName": "usr.Form.Form",
    "events": Array [
      Object {
        "name": "onSubmit",
        "targets": Array [
          Object {
            "events": Array [
              Object {
                "name": "resultData",
                "targets": Array [
                  Object {
                    "events": Array [
                      Object {
                        "name": "transformedData",
                        "targets": Array [
                          Object {
                            "events": Array [],
                            "props": Object {
                              "componentInstance": "pane",
                              "componentName": "usr.Pane.Pane",
                              "propertyName": "data",
                            },
                            "type": "component",
                          },
                          Object {
                            "events": Array [],
                            "props": Object {
                              "componentInstance": "form",
                              "componentName": "usr.Form.Form",
                              "propertyName": "formData",
                            },
                            "type": "component",
                          },
                          Object {
                            "events": Array [],
                            "props": Object {
                              "componentInstance": "pane",
                              "componentName": "usr.Pane.Pane",
                              "propertyName": "data3",
                            },
                            "type": "component",
                          },
                          Object {
                            "events": Array [],
                            "props": Object {
                              "functionName": "usr.api.myFunctions.showTransformedData",
                            },
                            "type": "userFunction",
                          },
                        ],
                      },
                    ],
                    "props": Object {
                      "functionName": "usr.api.myFunctions.transformData",
                    },
                    "type": "userFunction",
                  },
                ],
              },
              Object {
                "name": "success",
                "targets": Array [
                  Object {
                    "events": Array [],
                    "props": Object {
                      "componentInstance": "pane",
                      "componentName": "usr.Pane.Pane",
                      "propertyName": "data",
                    },
                    "type": "component",
                  },
                  Object {
                    "props": Object {
                      "componentInstance": "pane",
                      "componentName": "usr.Pane.Pane",
                      "propertyName": "data4",
                    },
                    "type": "component",
                  },
                  Object {
                    "events": Array [],
                    "props": Object {
                      "componentInstance": "pane",
                      "componentName": "usr.Pane.Pane",
                      "propertyName": "data2",
                    },
                    "type": "component",
                  },
                ],
              },
              Object {
                "name": "error",
                "targets": Array [
                  Object {
                    "props": Object {
                      "componentInstance": "errorPane",
                      "componentName": "usr.ErrorPane.ErrorPane",
                      "propertyName": "error",
                    },
                    "type": "component",
                  },
                  Object {
                    "props": Object {
                      "componentInstance": "pane",
                      "componentName": "usr.Pane.Pane",
                      "propertyName": "data2",
                    },
                    "type": "component",
                  },
                ],
              },
            ],
            "props": Object {
              "functionName": "usr.api.myFunctions.function1",
            },
            "type": "userFunction",
          },
        ],
      },
    ],
  },
  "usr.Pane.Pane_pane": Object {
    "componentInstance": "pane",
    "componentName": "usr.Pane.Pane",
    "events": Array [
      Object {
        "name": "onClick",
        "targets": Array [
          Object {
            "events": Array [
              Object {
                "name": "transformedData",
                "targets": Array [
                  Object {
                    "props": Object {
                      "componentInstance": "form",
                      "componentName": "usr.Form.Form",
                      "propertyName": "formData",
                    },
                    "type": "component",
                  },
                ],
              },
            ],
            "props": Object {
              "functionName": "usr.api.myFunctions.transformData",
            },
            "type": "userFunction",
          },
        ],
      },
    ],
    "propertyName": "data",
  },
}
`;

exports[`Sequences merge sequences 2`] = `
Object {
  "usr.ErrorPane.ErrorPane_errorPane": Object {
    "error": Object {
      "forwardPath": undefined,
    },
  },
  "usr.Form.Form_form": Object {
    "formData": Object {
      "forwardPath": undefined,
    },
  },
  "usr.Pane.Pane_pane": Object {
    "data": Object {
      "forwardPath": undefined,
    },
    "data2": Object {
      "forwardPath": undefined,
    },
    "data3": Object {
      "forwardPath": undefined,
    },
    "data4": Object {
      "forwardPath": undefined,
    },
  },
}
`;
